File: python/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if not l1:
            return l2
        if not l2:
            return l1
        dummyhead = ListNode(0)
        ptr = dummyhead
        p1 = l1
        p2 = l2
        carry, val = 0, 0
        while p1 or p2:
            val1 = p1.val if p1 else 0
            val2 = p2.val if p2 else 0
            val = val1 + val2 + carry
            carry = 1 if val >= 10 else 0
            val %= 10
            ptr.next = ListNode(val)
            ptr = ptr.next
            p1 = p1.next if p1 else None
            p2 = p2.next if p2 else None
        if carry:
            ptr.next = ListNode(carry)
        return dummyhead.next

File: cpp/1-50/2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {      
        int carry = 0;
        ListNode *head = new ListNode(0);
        head->val = 0;
        ListNode *ptr = head;
        while(l1 || l2|| carry)
        {
            int l1val = l1?l1->val:0;
            int l2val = l2?l2->val:0;
            int temp = l1val + l2val + carry;
            ptr->next = new ListNode(temp%10);
            ptr = ptr->next;
            carry = temp/10;
            l1 = l1?l1->next:NULL;
            l2 = l2?l2->next:NULL;
        }
        return head->next;
    }
};